src/phantom_token.c

```diff
@@ -265,63 +322,79 @@ static ngx_int_t handler(ngx_http_request_t *request)
                 {
-                    request->headers_in.accept->value = module_context->original_accept_header;
+                    if (headers_more_set_header_in(request, ACCEPT_HEADER_NAME, module_context->original_accept_header, &request->headers_in.accept) != NGX_OK)
+                    {
+                        utils_log_upstream_set_header_error(request, ACCEPT_HEADER_NAME);
+                        return utils_write_error_response(request, NGX_HTTP_INTERNAL_SERVER_ERROR, module_location_config);
+                    }
                 }
 
                 return NGX_OK;
             }
             else if (module_context->status == NGX_HTTP_NO_CONTENT)
             {
-                return set_www_authenticate_header(request, module_location_config, NULL);
+                return utils_set_www_authenticate_header(request, module_location_config, NULL);
             }
             else if (module_context->status == NGX_HTTP_SERVICE_UNAVAILABLE)
             {
-                return write_error_response(request, NGX_HTTP_SERVICE_UNAVAILABLE, module_location_config);
+                return utils_write_error_response(request, NGX_HTTP_SERVICE_UNAVAILABLE, module_location_config);
             }
-            else if (module_context->status >= NGX_HTTP_INTERNAL_SERVER_ERROR || module_context->status == NGX_HTTP_NOT_FOUND
-                || module_context->status == NGX_HTTP_UNAUTHORIZED || module_context->status == NGX_HTTP_FORBIDDEN)
+            else if (module_context->status >= NGX_HTTP_INTERNAL_SERVER_ERROR ||
+                     module_context->status == NGX_HTTP_NOT_FOUND ||
+                     module_context->status == NGX_HTTP_UNAUTHORIZED ||
+                     module_context->status == NGX_HTTP_FORBIDDEN)
             {
-                return write_error_response(request, NGX_HTTP_BAD_GATEWAY, module_location_config);
+                return utils_write_error_response(request, NGX_HTTP_BAD_GATEWAY, module_location_config);
             }
 
-            return write_error_response(request, NGX_HTTP_INTERNAL_SERVER_ERROR, module_location_config);
+            return utils_write_error_response(request, NGX_HTTP_INTERNAL_SERVER_ERROR, module_location_config);
         }
 
         ngx_log_debug0(NGX_LOG_DEBUG_HTTP, request->connection->log, 0,
-                       "Called again without having received the response from Curity");
+            "Called again without having received the introspection response");
 
         return NGX_AGAIN;
     }
 
     // return unauthorized when no Authorization header is present
     if (!request->headers_in.authorization || request->headers_in.authorization->value.len <= 0)
     {
-        ngx_log_debug0(NGX_LOG_DEBUG_HTTP, request->connection->log, 0, "Authorization header not present");
-
-        return set_www_authenticate_header(request, module_location_config, NULL);
+        ngx_log_error(NGX_LOG_WARN, request->connection->log, 0, "Authorization header not found");
```

Comment:
Could WARN be a bit too high? A misbehaving client or probing could cause these warnings..

Same for the other occurrence just below.

Reply:
OK - reverted back to a debug message for missing, invalid or expired tokens where the caller gets a 401. Other events are an unexpected backend error and get logged.